### PR TITLE
remove rows through __delitem__

### LINKED ISSUE #5832: Deleting a Table's row does not work
While looking at #5772 I noticed this:
```python
In [2]: t = Table(data=[[1,2,3],[4,5,6]])

In [3]: t
Out[3]: 
<Table length=3>
 col0  col1
int64 int64
----- -----
    1     4
    2     5
    3     6

In [4]: del t[1]

In [5]: t
Out[5]: 
<Table length=3>
 col0  col1
int64 int64
----- -----
    1     4
    2     5
    3     6
```
So here `del t[1]` does nothing. It should at least raise an exception if this operation is not supported, but I think it makes sense to support the same options as getitem/setitem ? (Which should be easy to do with `remove_row(s)`).

### PR DESCRIPTION
**Pull request by janga1997**
Fixes #5832 
@mwcraig 
@saimn Do suggest any changes if required.

### CODE REVIEW

File: astropy/table/table.py [hunk 1]
```diff
@@ -1312,9 +1312,16 @@ def __setitem__(self, item, value):
     def __delitem__(self, item):
         if isinstance(item, six.string_types):
             self.remove_column(item)
-        elif isinstance(item, tuple):
+
+        elif isinstance(item, (six.integer_types, np.integer)):
```
Comment (mirca) [inline astropy/table/table.py]: You can use ``_is_int`` from ``astropy.io.fits.utils`` to check for integer types

Comment (janga1997) [inline astropy/table/table.py]: I will edit that.
Apart from that, is there any other problem?

**Comment by bsipocz** [comment]
@janga1997 - This definitely needs a test and a changelog entry.

**Comment by bsipocz** [comment]
@pllim - I've milestoned it as a bugfix (maybe it can go back to the LTS release?), but you've labeled #5832 as a feature request. Please feel free to change the milestone if needed.

**Comment by janga1997** [comment]
@bsipocz yes it does. But the changes I've made till now are good enough, I assume?

**Comment by taldcroft** [comment]
@janga1997 - preferred practice for a relatively simple fix like this is to have passing tests included before you make a PR.  That way we don't waste CI time.  If you want early comment on the code without tests you can commit with `[skip ci]` in the commit message.

**Comment by janga1997** [comment]
@taldcroft That's good to know. Will keep that in mind in the future.

**Comment by pllim** [comment]
@bsipocz , the issue seems like to me it was requesting a non-existent feature, so that was why I labeled it was feature request.  @taldcroft , was that a bug report instead?

**Comment by mirca** [comment]
@janga1997 Look that travis failures are legit. Watch out for the white spaces ;-)

**Comment by janga1997** [comment]
@mirca Rest all is good enough?

File: astropy/table/table.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 
 from .. import log
 from ..io import registry as io_registry
+from ..io.fits.util import  _is_int
```
Comment (mhvk) [inline astropy/table/table.py]: We should not be importing a private function from another module!
In principle, python has the `__index__` method to check whether something can be used as an index, so you could write
```
def _is_int(item):
    try:
        item.__index__()
    except Exception:
        return False
    else:
        return True
```
This works for numpy integers as well.

I would probably put this function in `utils.misc as `isint`

File: astropy/table/tests/test_table.py [hunk 3]
```diff
@@ -849,6 +849,27 @@ def test_remove_row_preserves_meta(self, table_types):
...
+        self._setup(table_types)
+        self.t.add_column(self.b)
+        self.t.add_column(self.c)
+        del self.t[[0, 2]]
+        assert self.t.colnames == ['a', 'b', 'c']
+        assert np.all(self.t['c'] == np.array([8]))
+
+    def test_delitem_row_fail(self, table_types):
+        self._setup(table_types)
+        with pytest.raises(IndexError):
```
Comment (mhvk) [inline astropy/table/tests/test_table.py]: For completeness, also add a test here where you put in something that is not an integer, e.g., `del self.t[1.]`

**Comment by mhvk** [comment]
@janga1997 - a bit slow review, but two comments (with the first being important). 

Others: does my suggestion for adding an `isint` function to `utils.misc` make sense? In principle, also fine to just define it locally.

File: astropy/table/table.py [hunk 4]
```diff
@@ -1312,9 +1313,16 @@ def __setitem__(self, item, value):
...
             self.remove_column(item)
-        elif isinstance(item, tuple):
+
+        elif _is_int(item):
+            self.remove_row(item)
+
+        elif isinstance(item, (list, tuple, np.ndarray)) and all(isinstance(x, six.string_types) for x in item):
             self.remove_columns(item)
 
+        elif isinstance(item, (list, np.ndarray)) and all(_is_int(x) for x in item):
```
Comment (taldcroft) [inline astropy/table/table.py]: For performance do:
```
elif isinstance(item, (list, np.ndarray)) and np.asarray(x).dtype.kind == 'i':
```
This does create a temporary intermediate array, but in this case I think that is better than dropping to a slow pure Python loop.  For a million-long `item` this method is about 8 times faster.

File: astropy/table/table.py [hunk 5]
```diff
@@ -1312,9 +1313,16 @@ def __setitem__(self, item, value):
...
+            self.remove_rows(item)
```
Comment (taldcroft) [inline astropy/table/table.py]: Should be a final `else` and raise an exception that the `item` was incorrect.

File: astropy/table/table.py [hunk 6]
```diff
@@ -1312,9 +1313,16 @@ def __setitem__(self, item, value):
     def __delitem__(self, item):
         if isinstance(item, six.string_types):
             self.remove_column(item)
-        elif isinstance(item, tuple):
+
+        elif _is_int(item):
+            self.remove_row(item)
+
+        elif isinstance(item, (list, tuple, np.ndarray)) and all(isinstance(x, six.string_types) for x in item):
```
Comment (taldcroft) [inline astropy/table/table.py]: Should also cover (and test) the case of a slice.

File: astropy/table/tests/test_table.py [hunk 7]
```diff
@@ -849,6 +849,27 @@ def test_remove_row_preserves_meta(self, table_types):
...
+        self.t.add_column(self.c)
+        del self.t[1]
+        assert self.t.colnames == ['a', 'b', 'c']
+        assert np.all(self.t['a'] == np.array([1, 3]))
+
+    def test_delitem_row_list(self, table_types):
+        self._setup(table_types)
+        self.t.add_column(self.b)
+        self.t.add_column(self.c)
+        del self.t[[0, 2]]
```
Comment (taldcroft) [inline astropy/table/tests/test_table.py]: Test an `ndarray` like this.

File: astropy/table/table.py (hunk 2 above)
Comment (janga1997) [inline astropy/table/table.py]: So you want me to write a function `isint` as you have described above in `utils.misc`?
@taldcroft Any opinion?

File: astropy/table/tests/test_table.py [hunk 8]
```diff
@@ -849,6 +849,41 @@ def test_remove_row_preserves_meta(self, table_types):
...
+        self.t.add_column(self.c)
+        del self.t[idx]
+        assert self.t.colnames == ['a', 'b', 'c']
+        assert np.all(self.t['c'] == np.array([8]))
+
+    def test_delitem_row_slice(self, table_types):
+        self._setup(table_types)
+        self.t.add_column(self.b)
+        self.t.add_column(self.c)
+        del self.t[slice(0, 2, 1)]
```
Comment (taldcroft) [inline astropy/table/tests/test_table.py]: This should be expressed in the normal syntax `del self.t[0:2]`.

File: astropy/table/table.py (hunk 2 above)
Comment (taldcroft) [inline astropy/table/table.py]: There is no need for the `isint` function at this point with the `np.asarray` solution you have implemented.

Comment (taldcroft) [inline astropy/table/table.py]: Also, remove the `_is_int` import above.

File: astropy/table/table.py [hunk 9]
```diff
@@ -1312,9 +1313,22 @@ def __setitem__(self, item, value):
...
             self.remove_columns(item)
 
+        elif isinstance(item, (list, np.ndarray)) and np.asarray(item).dtype.kind == 'i':
+            self.remove_rows(item)
+
+        elif isinstance(item, slice):
+            self.remove_rows(item)
+
+        else:
+            raise IndexError('Illegal key or index value')
```
Comment (taldcroft) [inline astropy/table/table.py]: Don't capitalize `Illegal` (following precedent of Python exception messages).

**Comment by janga1997** [comment]
@taldcroft But we need the `_is_int` function to check for single integer as index.
What do we do about that?

**Comment by astrofrog** [comment]
@taldcroft - was this almost ready to merge?

**Comment by taldcroft** [comment]
It needs a little cleanup (perhaps by a maintainer), but this is worthwhile and I think we should re-open.

**Comment by saimn** [comment]
I rebased the branch and removed the use of `_is_int` (using a code similar to `__getitem__` seems good enough for now ?).

**Comment by taldcroft** [comment]
And thanks @janga1997 for getting most of the way on this.

**Pull request merged**